File: MoveAllZerors_day_11/zeros.js
// Brute force method

// Time: O(n)
// Space: O(n)

function movesZeros(arr){
    let result = []
    let zeroCount =0

    for(num of arr){
        if(num === 0){
            zeroCount++
        }else{
            result.push(num)
        }
    }

    while(num > 0){
        result.push(0)
        zeroCount--
    }

    return result
}


// Optimized way

// Time: O(n)
// Space: O(1)


function movesZerosToEnd(arr){
    let insertPos = 0

    for(let i=0; i<arr.length;i++){
        if(arr[i]!= 0){
            arr[insertPos] = arr[i]
            insertPos++
        }
    }

    while(insertPos <arr.length){
        arr[insertPos] = 0
        insertPos--
    }

    return arr
}


// swap method
// Time: O(n)
// Space: O(1)

function moveZerosArr(arr) {
    let insertPos = 0;

    for (let i = 0; i < arr.length; i++) {
        if (arr[i] !== 0) {
            [arr[insertPos], arr[i]] = [arr[i], arr[insertPos]];
            insertPos++;
        }
    }

    return arr;
}

File: MoveAllZerors_day_11/zeros.py
# Brute force method

# Time: O(n)
# Space: O(n)

def movesZeros(arr):
    result = []
    zeroCount = 0

    for num in arr:
       if(num == 0):
          zeroCount += 1
       else:
          result.append(num)

    while zeroCount > 0:
       result.append(0)
       zeroCount -= 1
      
    return result



# Optimized 
# Time: O(n)
# Space: O(n)

def movesZerosToEnd(arr):

    insert_pos = 0

    for i in range (len(arr)):
       if(arr[i] != 0):
          arr[insert_pos] = arr[i]
          insert_pos += 1

       while(insert_pos < len(arr)):
          arr[insert_pos] = 0
          insert_pos -= 1

       return arr
 
# Swap
# Time: O(n)
# Space: O(n)

def move_zeros(arr):
    insert_pos = 0

    for i in range(len(arr)):
        if arr[i] != 0:
            arr[insert_pos], arr[i] = arr[i], arr[insert_pos]
            insert_pos += 1

    return arr